espnet2/bin/s2t_inference.py

```diff
@@ -0,0 +1,1040 @@
+        Returns:
+            text, token, token_int, hyp
+
+        """
+        assert check_argument_types()
+
+        # Prepare hyp_primer
+        if text_prev is not None:
+            if isinstance(text_prev, str):
+                text_prev = self.converter.tokens2ids(
+                    self.tokenizer.text2tokens(text_prev)
+                )
+            else:
+                text_prev = text_prev.tolist()
+
+            # Check if text_prev is valid
+            if self.s2t_model.na in text_prev:
+                text_prev = None
+
+        if text_prev is not None:
+            hyp_primer = (
+                [self.s2t_model.sop]
+                + text_prev
+                + [self.s2t_model.sos, self.category_id, self.task_id]
+            )
+        else:
+            hyp_primer = [self.s2t_model.sos, self.category_id, self.task_id]
+        if self.time_id is not None:
+            hyp_primer.append(self.time_id)
+        self.beam_search.set_hyp_primer(hyp_primer)
+
+        # Preapre speech
+        if isinstance(speech, np.ndarray):
+            speech = torch.tensor(speech)
+
+        # Batchify input
+        # speech: (nsamples,) -> (1, nsamples)
+        speech = speech.unsqueeze(0).to(getattr(torch, self.dtype))
+        # lengths: (1,)
+        lengths = speech.new_full([1], dtype=torch.long, fill_value=speech.size(1))
+        batch = {"speech": speech, "speech_lengths": lengths}
+        logging.info("speech length: " + str(speech.size(1)))
+
+        # a. To device
+        batch = to_device(batch, device=self.device)
+
+        # b. Forward Encoder
+        enc, enc_olens = self.s2t_model.encode(**batch)
+
+        intermediate_outs = None
+        if isinstance(enc, tuple):
+            enc, intermediate_outs = enc
+
+        assert len(enc) == 1, len(enc)
+
+        # c. Pass the encoder result to the beam search
+        results = self._decode_single_sample(enc[0])
+
+        # Encoder intermediate CTC predictions
+        if intermediate_outs is not None:
+            encoder_interctc_res = self._decode_interctc(intermediate_outs)
+            results = (results, encoder_interctc_res)
+
+        assert check_return_type(results)
+
+        return results
+
+    def _decode_interctc(
+        self, intermediate_outs: List[Tuple[int, torch.Tensor]]
+    ) -> Dict[int, List[str]]:
+        assert check_argument_types()
+
+        exclude_ids = [self.s2t_model.blank_id, self.s2t_model.sos, self.s2t_model.eos]
+        res = {}
+        token_list = self.beam_search.token_list
+
+        for layer_idx, encoder_out in intermediate_outs:
+            y = self.s2t_model.ctc.argmax(encoder_out)[0]  # batch_size = 1
+            y = [x[0] for x in groupby(y) if x[0] not in exclude_ids]
+            y = [token_list[x] for x in y]
+
+            res[layer_idx] = y
+
+        return res
+
+    def _decode_single_sample(self, enc: torch.Tensor):
+        if self.beam_search_transducer:
```

Comment:
The inference seems not be finished, will check later.

Reply:
Inference (utterance-level or long-form) is finished.